engine/player/unique_gear_shadowlands.cpp

```diff
@@ -4990,6 +5038,19 @@ void register_target_data_initializers( sim_t& sim )
       td->debuff.volatile_satchel = make_buff( *td, "volatile_satchel" )->set_quiet( true );
   } );
 
+  // Bells of the Endless Feast
+  sim.register_target_data_initializer( []( actor_target_data_t* td ) {
+    if ( unique_gear::find_special_effect( td->source, 367336 ) )
+    {
+      assert( !td->debuff.scent_of_souls );
+
+      td->debuff.scent_of_souls = make_buff<buff_t>( *td, "scent_of_souls", td->source->find_spell( 368585 ) );
+      td->debuff.scent_of_souls->reset();
```

Comment:
It seemed that it triggers on the tick rather than immediately after reaching max stacks. The only scenario where this is relevant is if you would proc the trinket directly after getting the max stacks. The current model would reset the debuff and award you with new stacks. 

With the actual in-game behavior, there is a possibility that you are still on max stacks since the debuff hasn't ticked yet by the time you proc the trinket again, meaning the debuff hasn't reset yet, and you would lose new stacks. 

I'm not really sure we need to model it this way because of it.

Reply:
Hm, in that case I'd model the in-game behavior tbh. Considering it should be easy.

Proc action should still be created in the special effect initialization function (because actions have to be created before the simulation starts) but it can be dropped from the custom `dbc_proc_callback_t` with only debuff triggering left there.

Then in the target data initializer just find the proc action and slap a tick callback on the debuff triggering it. Smth like
```cpp
auto damage = td->source->find_action( "rabid_devourer_chomp" );
assert( damage );
[...]
td->debuff.scent_of_souls->set_tick_callback( [ damage ]( buff_t* b, int, timespan_t ){
    if ( b->at_max_stacks() ) {
      damage->execute_on_target( b->player );
      b->expire( 1_ms );
    }
  } );
```
`1_ms` in the expire is needed because we can't directly expire buffs from callbacks, so there will be a 1ms window where you can lose stacks, but eh.

I *think* this should work and properly model in-game behaviour.

Reply:
@nuoHep hmm, I've implemented it as suggested, but I've noticed some issues with the `at_max_stacks` method. Once the debuff reaches max stacks method continues returning true, even tho debuff was expired. I can see the stack fall off, but the damage portion is triggered on each tick never the less.

![](https://i.imgur.com/HjNUN43.png)
![](https://i.imgur.com/f1TQp6b.png)

As you can see, the uptime on 100 stacks is 5%, but the damage is triggered over 600 times.

Reply:
Hm, interesting.

Can you please rebase it once more so we can merge this as is. I'll try to look at it later why this happens.

Reply:
I've rebased it. I also saw that you whitelisted the damage spell id and removed the commit with the generated spell data since it already contains everything.